File: numerics/src/hmm_em.py
"""EM (Baum-Welch) for a univariate Gaussian HMM, with scaled forward-backward.

Design choices
--------------
- Scaled forward-backward (Rabiner-style) to avoid underflow.
- Multiple random initializations; the best log-likelihood is kept.
- A minimum standard deviation `sigma_min` prevents degenerate components.
- Fitted states are reordered by increasing mean before being returned.
- The EM trace is checked for (numerical) monotonicity of the log-likelihood.

Pure NumPy: no external HMM library is required.
"""
from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

SIGMA_MIN = 0.05  # minimal emission standard deviation


@dataclass
class HMMFit:
    """Result of one EM fit (best over random starts), states sorted by mean."""
    K: int
    loglik: float
    pi: np.ndarray          # (K,) initial distribution
    A: np.ndarray           # (K, K) transition matrix
    means: np.ndarray       # (K,)
    sds: np.ndarray         # (K,)
    gamma: np.ndarray       # (T, K) posterior state probabilities
    viterbi: np.ndarray     # (T,) most likely state path (0-based)
    n_iter: int
    loglik_trace: list = field(default_factory=list)

    def n_params(self) -> int:
        return (self.K - 1) + self.K * (self.K - 1) + 2 * self.K

    def bic(self, T: int) -> float:
        return -2.0 * self.loglik + self.n_params() * np.log(T)

    def icl(self, T: int) -> float:
        """ICL = BIC + 2 * posterior entropy."""
        g = np.clip(self.gamma, 1e-300, 1.0)
        entropy = -float(np.sum(self.gamma * np.log(g)))
        return self.bic(T) + 2.0 * entropy


def _emission_probs(y, means, sds):
    var = sds ** 2
    B = np.exp(-0.5 * (y[:, None] - means[None, :]) ** 2 / var[None, :])
    B /= np.sqrt(2.0 * np.pi * var[None, :])
    return np.maximum(B, 1e-300)


def _fb_core(B, pi, A):
    """Scaled forward-backward recursions on the emission matrix B (T, K)."""
    T, K = B.shape
    alpha = np.zeros((T, K))
    scale = np.zeros(T)
    alpha[0] = pi * B[0]
    scale[0] = alpha[0].sum()
    alpha[0] /= scale[0]
    for t in range(1, T):
        alpha[t] = (alpha[t - 1] @ A) * B[t]
        scale[t] = alpha[t].sum()
        alpha[t] /= scale[t]
    beta = np.zeros((T, K))
    beta[T - 1] = 1.0
    for t in range(T - 2, -1, -1):
        beta[t] = A @ (B[t + 1] * beta[t + 1])
        beta[t] /= scale[t + 1]
    gamma = alpha * beta
    gamma /= gamma.sum(axis=1, keepdims=True)
    obs = B[1:] * beta[1:]
    xi = alpha[:-1, :, None] * A[None, :, :] * obs[:, None, :]
    xi_sum = (xi / xi.sum(axis=(1, 2), keepdims=True)).sum(axis=0)
    loglik = np.sum(np.log(scale))
    return loglik, gamma, xi_sum


def forward_backward(y, pi, A, means, sds):
    """Scaled forward-backward. Returns (loglik, gamma, xi_sum)."""
    B = _emission_probs(y, means, sds)
    loglik, gamma, xi_sum = _fb_core(B, pi, A)
    return float(loglik), gamma, xi_sum


def viterbi_path(y, pi, A, means, sds):
    """Viterbi decoding in log-space."""
    T, K = len(y), len(pi)
    logB = np.log(_emission_probs(y, means, sds))
    logA = np.log(np.maximum(A, 1e-300))
    delta = np.zeros((T, K))
    psi = np.zeros((T, K), dtype=int)
    delta[0] = np.log(np.maximum(pi, 1e-300)) + logB[0]
    for t in range(1, T):
        cand = delta[t - 1][:, None] + logA
        psi[t] = cand.argmax(axis=0)
        delta[t] = cand.max(axis=0) + logB[t]
    path = np.zeros(T, dtype=int)
    path[-1] = int(delta[-1].argmax())
    for t in range(T - 2, -1, -1):
        path[t] = psi[t + 1, path[t + 1]]
    return path


def _em_single(y, K, rng, max_iter, tol, sigma_min):
    """One EM run from a random initialization. Returns dict or None on failure."""
    T = len(y)
    y_sd = float(np.std(y))
    # Random-but-sensible initialization: spread means over quantiles, jittered.
    means = np.quantile(y, np.linspace(0.1, 0.9, K)) + rng.normal(0.0, 0.25 * y_sd, K)
    sds = np.full(K, max(2 * sigma_min, y_sd / 2.0)) * rng.uniform(0.6, 1.4, K)
    pi = rng.dirichlet(np.ones(K))
    A = np.vstack([rng.dirichlet(np.ones(K) + 6.0 * np.eye(K)[i]) for i in range(K)])

    trace = []
    prev = -np.inf
    for it in range(max_iter):
        loglik, gamma, xi_sum = forward_backward(y, pi, A, means, sds)
        trace.append(loglik)
        # Monotonicity check (tiny numerical tolerance allowed)
        if loglik < prev - 1e-6 * (1.0 + abs(prev)):
            raise RuntimeError(
                f"EM log-likelihood decreased at iteration {it}: {prev} -> {loglik}"
            )
        converged = (loglik - prev) < tol * (1.0 + abs(prev))
        prev = loglik
        # M-step
        weights = gamma.sum(axis=0) + 1e-12
        pi = gamma[0] + 1e-10
        pi /= pi.sum()
        A = xi_sum + 1e-10
        A /= A.sum(axis=1, keepdims=True)
        means = (gamma * y[:, None]).sum(axis=0) / weights
        var = (gamma * (y[:, None] - means[None, :]) ** 2).sum(axis=0) / weights
        sds = np.sqrt(np.maximum(var, sigma_min ** 2))
        if converged and it > 0:
            break
    loglik, gamma, _ = forward_backward(y, pi, A, means, sds)
    trace.append(loglik)
    return dict(loglik=loglik, pi=pi, A=A, means=means, sds=sds,
                gamma=gamma, n_iter=len(trace) - 1, trace=trace)


def fit_hmm(y, K, seed, n_starts=20, max_iter=300, tol=1e-6,
            sigma_min=SIGMA_MIN) -> HMMFit:
    """Fit a K-state univariate Gaussian HMM by EM with multiple random starts.

    States in the returned fit are ordered by increasing mean.
    """
    rng = np.random.default_rng(seed)
    best = None
    for _ in range(n_starts):
        res = _em_single(y, K, rng, max_iter, tol, sigma_min)
        if best is None or res["loglik"] > best["loglik"]:
            best = res
    order = np.argsort(best["means"])
    pi = best["pi"][order]
    A = best["A"][np.ix_(order, order)]
    means = best["means"][order]
    sds = best["sds"][order]
    gamma = best["gamma"][:, order]
    vit = viterbi_path(y, pi, A, means, sds)
    return HMMFit(K=K, loglik=best["loglik"], pi=pi, A=A, means=means,
                  sds=sds, gamma=gamma, viterbi=vit, n_iter=best["n_iter"],
                  loglik_trace=best["trace"])


def validate_fit(fit: HMMFit, T: int) -> None:
    """Assertions on a fitted model; raises AssertionError on failure."""
    assert fit.gamma.shape == (T, fit.K), "gamma has wrong shape"
    assert np.allclose(fit.A.sum(axis=1), 1.0, atol=1e-8), "rows of A must sum to 1"
    assert np.allclose(fit.pi.sum(), 1.0, atol=1e-8), "pi must sum to 1"
    assert np.all(fit.sds > 0), "estimated sds must be positive"
    assert np.allclose(fit.gamma.sum(axis=1), 1.0, atol=1e-8), \
        "posterior probabilities must sum to 1 at each time"
    assert np.all(np.isfinite(fit.gamma)), "gamma contains NaN/Inf"
    assert np.all(np.isfinite([fit.loglik, *fit.means, *fit.sds])), \
        "parameters contain NaN/Inf"
    tr = np.array(fit.loglik_trace)
    assert np.all(np.diff(tr) >= -1e-6 * (1.0 + np.abs(tr[:-1]))), \
        "EM log-likelihood is not non-decreasing (beyond tolerance)"

File: numerics/R/hmm_em.R
# Expectation-maximization for a univariate Gaussian hidden Markov model.
#
# The implementation uses scaled forward-backward recursions, multiple random
# starts, a lower bound on emission standard deviations, and Viterbi decoding.
# No external HMM package is required.

SIGMA_MIN <- 0.05

normalize_probability <- function(x) {
  total <- sum(x)
  if (!is.finite(total) || total <= 0) {
    stop("Cannot normalize a non-positive probability vector.", call. = FALSE)
  }
  x / total
}

normalize_rows <- function(x) {
  totals <- rowSums(x)
  if (any(!is.finite(totals)) || any(totals <= 0)) {
    stop("Cannot normalize a matrix containing a non-positive row.", call. = FALSE)
  }
  x / totals
}

rdirichlet1 <- function(alpha) {
  normalize_probability(stats::rgamma(length(alpha), shape = alpha, rate = 1))
}

sequence_indices <- function(T, lengths = NULL) {
  if (is.null(lengths)) {
    return(list(seq_len(T)))
  }
  lengths <- as.integer(lengths)
  if (length(lengths) == 0L || anyNA(lengths) || any(lengths <= 0L)) {
    stop("lengths must contain positive sequence lengths.", call. = FALSE)
  }
  if (sum(lengths) != T) {
    stop("Sequence lengths must sum to the number of observations.", call. = FALSE)
  }
  ends <- cumsum(lengths)
  starts <- c(1L, head(ends, -1L) + 1L)
  Map(seq.int, starts, ends)
}

gaussian_emission_probabilities <- function(y, means, sds) {
  output <- vapply(
    seq_along(means),
    function(j) stats::dnorm(y, mean = means[[j]], sd = sds[[j]]),
    numeric(length(y))
  )
  output <- matrix(output, nrow = length(y), ncol = length(means))
  pmax(output, 1e-300)
}

forward_backward_core <- function(emission, initial, transition) {
  T <- nrow(emission)
  K <- ncol(emission)
  alpha <- matrix(0, nrow = T, ncol = K)
  scale <- numeric(T)

  alpha[1L, ] <- initial * emission[1L, ]
  scale[[1L]] <- sum(alpha[1L, ])
  alpha[1L, ] <- alpha[1L, ] / scale[[1L]]

  if (T > 1L) {
    for (time in 2:T) {
      alpha[time, ] <- as.numeric(alpha[time - 1L, ] %*% transition) *
        emission[time, ]
      scale[[time]] <- sum(alpha[time, ])
      alpha[time, ] <- alpha[time, ] / scale[[time]]
    }
  }

  beta <- matrix(0, nrow = T, ncol = K)
  beta[T, ] <- 1
  if (T > 1L) {
    for (time in seq.int(T - 1L, 1L)) {
      beta[time, ] <- as.numeric(
        transition %*% (emission[time + 1L, ] * beta[time + 1L, ])
      ) / scale[[time + 1L]]
    }
  }

  posterior <- alpha * beta
  posterior <- posterior / rowSums(posterior)
  transition_sum <- matrix(0, nrow = K, ncol = K)
  if (T > 1L) {
    for (time in seq_len(T - 1L)) {
      xi <- transition * outer(
        alpha[time, ],
        emission[time + 1L, ] * beta[time + 1L, ]
      )
      transition_sum <- transition_sum + xi / sum(xi)
    }
  }

  list(
    log_likelihood = sum(log(scale)),
    posterior = posterior,
    transition_sum = transition_sum
  )
}

forward_backward <- function(y, initial, transition, means, sds, lengths = NULL) {
  emission <- gaussian_emission_probabilities(y, means, sds)
  posterior <- matrix(0, nrow = length(y), ncol = length(initial))
  transition_sum <- matrix(0, nrow = length(initial), ncol = length(initial))
  log_likelihood <- 0

  for (indices in sequence_indices(length(y), lengths)) {
    result <- forward_backward_core(emission[indices, , drop = FALSE], initial, transition)
    posterior[indices, ] <- result$posterior
    transition_sum <- transition_sum + result$transition_sum
    log_likelihood <- log_likelihood + result$log_likelihood
  }

  list(
    log_likelihood = log_likelihood,
    posterior = posterior,
    transition_sum = transition_sum
  )
}

viterbi_single_gaussian <- function(y, initial, transition, means, sds) {
  T <- length(y)
  K <- length(initial)
  log_emission <- log(gaussian_emission_probabilities(y, means, sds))
  log_transition <- log(pmax(transition, 1e-300))
  delta <- matrix(0, nrow = T, ncol = K)
  back_pointer <- matrix(1L, nrow = T, ncol = K)
  delta[1L, ] <- log(pmax(initial, 1e-300)) + log_emission[1L, ]

  if (T > 1L) {
    for (time in 2:T) {
      for (destination in seq_len(K)) {
        candidates <- delta[time - 1L, ] + log_transition[, destination]
        back_pointer[time, destination] <- which.max(candidates)
        delta[time, destination] <- max(candidates) + log_emission[time, destination]
      }
    }
  }

  path <- integer(T)
  path[[T]] <- which.max(delta[T, ])
  if (T > 1L) {
    for (time in seq.int(T - 1L, 1L)) {
      path[[time]] <- back_pointer[time + 1L, path[[time + 1L]]]
    }
  }
  path
}

viterbi_gaussian <- function(y, initial, transition, means, sds, lengths = NULL) {
  path <- integer(length(y))
  for (indices in sequence_indices(length(y), lengths)) {
    path[indices] <- viterbi_single_gaussian(
      y[indices], initial, transition, means, sds
    )
  }
  path
}

gaussian_em_loop <- function(
    y, initial, transition, means, sds, max_iter, tolerance, sigma_min,
    lengths = NULL) {
  trace <- numeric()
  previous <- -Inf
  indices <- sequence_indices(length(y), lengths)
  starts <- vapply(indices, function(index) index[[1L]], integer(1))

  for (iteration in seq_len(max_iter)) {
    fb <- forward_backward(y, initial, transition, means, sds, lengths)
    log_likelihood <- fb$log_likelihood
    trace <- c(trace, log_likelihood)

    if (is.finite(previous) &&
        log_likelihood < previous - 1e-6 * (1 + abs(previous))) {
      stop(
        sprintf(
          "EM log-likelihood decreased at iteration %d: %.12f to %.12f.",
          iteration, previous, log_likelihood
        ),
        call. = FALSE
      )
    }
    converged <- is.finite(previous) &&
      (log_likelihood - previous) < tolerance * (1 + abs(previous))
    previous <- log_likelihood

    posterior <- fb$posterior
    weights <- colSums(posterior) + 1e-12
    initial <- normalize_probability(colMeans(posterior[starts, , drop = FALSE]) + 1e-10)
    transition <- normalize_rows(fb$transition_sum + 1e-10)
    means <- colSums(posterior * y) / weights
    centered <- sweep(matrix(y, nrow = length(y), ncol = length(means)), 2L, means)
    variances <- colSums(posterior * centered^2) / weights
    sds <- sqrt(pmax(variances, sigma_min^2))

    if (converged && iteration > 1L) {
      break
    }
  }

  fb <- forward_backward(y, initial, transition, means, sds, lengths)
  trace <- c(trace, fb$log_likelihood)
  list(
    log_likelihood = fb$log_likelihood,
    initial = initial,
    transition = transition,
    means = means,
    sds = sds,
    posterior = fb$posterior,
    n_iter = length(trace) - 1L,
    log_likelihood_trace = trace
  )
}

gaussian_em_start <- function(
    y, K, max_iter, tolerance, sigma_min, lengths = NULL) {
  y_sd <- stats::sd(y)
  means <- as.numeric(stats::quantile(y, seq(0.1, 0.9, length.out = K), names = FALSE)) +
    stats::rnorm(K, 0, 0.25 * y_sd)
  sds <- rep(max(2 * sigma_min, y_sd / 2), K) * stats::runif(K, 0.6, 1.4)
  initial <- rdirichlet1(rep(1, K))
  transition <- t(vapply(
    seq_len(K),
    function(state) rdirichlet1(rep(1, K) + 6 * (seq_len(K) == state)),
    numeric(K)
  ))
  gaussian_em_loop(
    y, initial, transition, means, sds, max_iter, tolerance, sigma_min, lengths
  )
}

gaussian_n_parameters <- function(fit) {
  (fit$K - 1L) + fit$K * (fit$K - 1L) + 2L * fit$K
}

fit_bic <- function(fit, T) {
  -2 * fit$log_likelihood + fit$n_parameters * log(T)
}

fit_icl <- function(fit, T) {
  posterior <- pmax(fit$posterior, 1e-300)
  entropy <- -sum(fit$posterior * log(posterior))
  fit_bic(fit, T) + 2 * entropy
}

fit_gaussian_hmm <- function(
    y, K, seed, n_starts = 20L, max_iter = 300L, tolerance = 1e-6,
    sigma_min = SIGMA_MIN, screen_iter = NULL, refine_top = NULL,
    screen_tolerance = 1e-4, lengths = NULL) {
  stopifnot(length(K) == 1L, K >= 1L, n_starts >= 1L)
  set.seed(seed)

  if (!is.null(screen_iter) && !is.null(refine_top) && refine_top < n_starts) {
    screened <- lapply(seq_len(n_starts), function(unused) {
      gaussian_em_start(
        y, K, as.integer(screen_iter), screen_tolerance, sigma_min, lengths
      )
    })
    selected <- screened[order(
      vapply(screened, `[[`, numeric(1), "log_likelihood"),
      decreasing = TRUE
    )[seq_len(refine_top)]]
    candidates <- lapply(selected, function(result) {
      gaussian_em_loop(
        y, result$initial, result$transition, result$means, result$sds,
        max_iter, tolerance, sigma_min, lengths
      )
    })
  } else {
    candidates <- lapply(seq_len(n_starts), function(unused) {
      gaussian_em_start(y, K, max_iter, tolerance, sigma_min, lengths)
    })
  }

  best <- candidates[[which.max(vapply(
    candidates, `[[`, numeric(1), "log_likelihood"
  ))]]
  ordering <- order(best$means)
  fit <- list(
    model = "Gaussian HMM",
    K = as.integer(K),
    log_likelihood = best$log_likelihood,
    initial = best$initial[ordering],
    transition = best$transition[ordering, ordering, drop = FALSE],
    means = best$means[ordering],
    sds = best$sds[ordering],
    posterior = best$posterior[, ordering, drop = FALSE],
    n_iter = best$n_iter,
    log_likelihood_trace = best$log_likelihood_trace
  )
  fit$n_parameters <- gaussian_n_parameters(fit)
  fit$viterbi <- viterbi_gaussian(
    y, fit$initial, fit$transition, fit$means, fit$sds, lengths
  )
  class(fit) <- c("gaussian_hmm_fit", "hmm_fit")
  fit
}

validate_gaussian_fit <- function(fit, T) {
  stopifnot(
    inherits(fit, "gaussian_hmm_fit"),
    identical(dim(fit$posterior), c(as.integer(T), fit$K)),
    max(abs(rowSums(fit$transition) - 1)) < 1e-8,
    abs(sum(fit$initial) - 1) < 1e-8,
    all(fit$sds > 0),
    max(abs(rowSums(fit$posterior) - 1)) < 1e-8,
    all(is.finite(fit$posterior)),
    all(is.finite(c(fit$log_likelihood, fit$means, fit$sds)))
  )
  differences <- diff(fit$log_likelihood_trace)
  tolerance <- -1e-6 * (1 + abs(head(fit$log_likelihood_trace, -1L)))
  stopifnot(all(differences >= tolerance))
  invisible(TRUE)
}
